### DISCUSSION TITLE
Example for connections in access point mode?

### DISCUSSION THREAD

**Discussion by tommasoclini** [MaJerle/lwesp#131] [Q&A]
I'm having trouble with connections in AP mode, this is the "flow" of my code:
1. initialize LwESP library
2. set Wi-Fi AP mode
3. configure the AP
4. start a UDP connection
The problem occurs at step 4, if I define LWESP_CFG_MODE_STATION to 1(default), I get error 9, if I define it to 0, I get error 2.
Here's my code(I'm using a nucleo f411re with an esp32-wroom-32), am I doing something wrong?:

void Startinit_thread(void *argument) {
	/* USER CODE BEGIN Startinit_thread */
	{
		lwespr_t res;
		if ((res = lwesp_init(examples_common_lwesp_callback_func, 1)) != lwespOK) {
			printf("Failed initializing LwESP, error code: %i, stopping\r\n", res);
			Error_Handler();
		}

	lwesp_evt_register(lwesp_global_events_func);

	/* Enable access point only mode */
	if ((res = lwesp_set_wifi_mode(LWESP_MODE_AP, NULL, NULL, 1)) == lwespOK) {
			printf("ESP set to access-point-only mode\r\n");
	} else {
			printf("Problems setting ESP to access-point-only mode: %d\r\n", (int)res);
			Error_Handler();
	}

	/* Configure access point */
	res = lwesp_ap_set_config(AP_SSID, AP_PASSWD, 13, LWESP_ECN_WPA2_PSK, 5, 0, NULL, NULL, 1);
	if (res == lwespOK) {
			printf("Access point configured!\r\n");
	} else {
			printf("Cannot configure access point!\r\n");
			Error_Handler();
	}

	lwesp_ip_t ap_ip;

	res = lwesp_ap_getip(&ap_ip, NULL, NULL, NULL, NULL, 1);
	if (res != lwespOK){
		printf("Error(%d) while trying to retrieve ap ip information\r\n", res);
		Error_Handler();
	} else {
		utils_print_ip("AP IP address: ", &ap_ip, "");
	}

	res = lwesp_conn_start(NULL, LWESP_CONN_TYPE_UDP, REMOTE_HOST, 8888, NULL, conn_callback_func, 1);
	if (res != lwespOK){
		printf("Cannot start connection, error code: %d\r\n", res);
		Error_Handler();
	}

	}
	for (;;){
		osDelay(1000);
	}
	/* USER CODE END Startinit_thread */
}

lwespr_t lwesp_global_events_func(lwesp_evt_t *evt) {
	switch (lwesp_evt_get_type(evt)) {
	case LWESP_EVT_AP_CONNECTED_STA: {
		lwesp_mac_t *mac = lwesp_evt_ap_connected_sta_get_mac(evt);
		utils_print_mac("New station connected to access point with MAC address: ",
				mac, "\r\n");
		break;
	}
	case LWESP_EVT_AP_IP_STA: {
		lwesp_mac_t *mac = lwesp_evt_ap_ip_sta_get_mac(evt);
		lwesp_ip_t *ip = lwesp_evt_ap_ip_sta_get_ip(evt);

		utils_print_ip("IP ", ip, " assigned to station with MAC address: ");
		utils_print_mac(NULL, mac, "\r\n");
		break;
	}
	case LWESP_EVT_AP_DISCONNECTED_STA: {
		lwesp_mac_t *mac = lwesp_evt_ap_disconnected_sta_get_mac(evt);
		utils_print_mac("Station disconnected from access point with MAC address: ",
				mac, "\r\n");
		break;
	}
	default:
		break;
	}
	return lwespOK;
}

lwespr_t conn_callback_func(lwesp_evt_t* evt) {
    lwesp_conn_p conn;
    //lwespr_t res;
    uint8_t conn_num;

    conn = lwesp_conn_get_from_evt(evt);          /* Get connection handle from event */
    if (conn == NULL) {
        return lwespERR;
    }
    conn_num = lwesp_conn_getnum(conn);           /* Get connection number for identification */
    switch (lwesp_evt_get_type(evt)) {
        case LWESP_EVT_CONN_ACTIVE: {             /* Connection just active */
            printf("Connection %d active!\r\n", (int)conn_num);
            break;
        }
        case LWESP_EVT_CONN_CLOSE: {              /* Connection closed */
            if (lwesp_evt_conn_close_is_forced(evt)) {
                printf("Connection %d closed by client!\r\n", (int)conn_num);
            } else {
                printf("Connection %d closed by remote side!\r\n", (int)conn_num);
            }
            break;
        }
        case LWESP_EVT_CONN_SEND: {               /* Data send event */
            lwespr_t res = lwesp_evt_conn_send_get_result(evt);
            if (res == lwespOK) {
                printf("Data sent successfully on connection %d...waiting to receive data from remote side...\r\n", (int)conn_num);
            } else {
                printf("Error while sending data on connection %d!\r\n", (int)conn_num);
            }
            break;
        }
        case LWESP_EVT_CONN_RECV: {               /* Data received from remote side */
            lwesp_pbuf_p pbuf = lwesp_evt_conn_recv_get_buff(evt);
            lwesp_conn_recved(conn, pbuf);        /* Notify stack about received pbuf */
            size_t pbuf_size = lwesp_pbuf_length(pbuf, 1);
            printf("Received %d bytes on connection %d: %.*s\r\n", (int)pbuf_size, (int)conn_num, (int)pbuf_size, (char*)lwesp_pbuf_get_linear_addr(pbuf, 0, NULL));
            break;
        }
        case LWESP_EVT_CONN_ERROR: {              /* Error connecting to server */
            const char* host = lwesp_evt_conn_error_get_host(evt);
            lwesp_port_t port = lwesp_evt_conn_error_get_port(evt);
            printf("Error connecting to %s:%d\r\n", host, (int)port);
            break;
        }
        default:
            break;
    }
    return lwespOK;
}

**Comment by MaJerle** [comment]
You cannot start connection if you are not connected to a wifi access point.

**Reply by tommasoclini** [reply]
So, how do I start one when the esp32 is the access point itself? Or how do I listen for udp messages?

**Reply by MaJerle** [reply]
If you want to listen for a connection, you have to start a server mode. You can only start a connection when you are, as a station, connected to another access point.

**Reply by tommasoclini** [reply]
Thank you for your time and answer, next time I'll check the documentation first

**Reply by tommasoclini** [reply]
I still can't quite grasp how the whole connection/server/client system work, can you show me a basic example of listening for udp messages when the esp32 is the access point itself? Just which api functions I need to call, nothing fancy

**Reply by MaJerle** [reply]
1. be in AP mode
2. accept client to connect
3. make sure you have server enable
4. I'm not sure you can listen on UDP via AT commands. Check ESP-AT user manual first.

**Reply by tommasoclini** [reply]
I had written some code that set up an udp connection with the esp32(esp8266) as the access point, but I was sending commands manually.
I had enabled transparent transmission and it had worked just fine. I had used this command:
https://docs.espressif.com/projects/esp-at/en/latest/esp32/AT_Command_Set/TCP-IP_AT_Commands.html#cmd-start (go to the udp transmission section)

**Reply by MaJerle** [reply]
Can you list the AT commands that have been used? I am not sure I can answer otherwise.

**Reply by tommasoclini** [reply]
I'm using termite, so the commands I send are mixed with the responses, the commands I've sent are bold:
ready
**AT+CWMODE=2**
AT+CWMODE=2

OK
**AT+CWSAP="AP_SSID","",5,0**
AT+CWSAP="AP_SSID","",5,0

OK
+STA_CONNECTED:"e8:84:a5:8e:60:8e"
+DIST_STA_IP:"e8:84:a5:8e:60:8e","192.168.4.2"
+STA_DISCONNECTED:"e8:84:a5:8e:60:8e"
+STA_CONNECTED:"e8:84:a5:8e:60:8e"
+DIST_STA_IP:"e8:84:a5:8e:60:8e","192.168.4.2"
**AT+CIPSTART="UDP","192.168.4.1",8888,8888**
AT+CIPSTART="UDP","192.168.4.1",8888,8888
+LINK_CONN:0,0,"UDP",0,"192.168.4.1",8888,8888

OK

+IPD,9:MESSAGE


+IPD,11:MESSAGE_1


+IPD,11:MESSAGE_2
**AT+CIPMODE=1**
AT+CIPMODE=1

OK
MESSAGE_3
MESSAGE_4
MESSAGE_5

**Reply by MaJerle** [reply]
So in your case, the sequence you have here is:

- You go to AP mode -> this you have done with lwesp too
- you have to wait for a station to connect. There is a `LWESP_EVT_AP_IP_STA` event when station receives an IP from AP.
- Only after you receive this event, you can do lwesp_conn_start, with the IP address of the station connected.
- And it should work.

**Reply by tommasoclini** [reply]
When sending commands manually, it also works with a dummy IP address, I don't think that's the problem:
**AT+CWMODE=2**
AT+CWMODE=2

OK
**AT+CWSAP="AP_SSID","",5,0**
AT+CWSAP="AP_SSID","",5,0

OK
**AT+CIPSTART="UDP","0.0.0.0",8888,8888**
AT+CIPSTART="UDP","0.0.0.0",8888,8888
+LINK_CONN:0,0,"UDP",0,"0.0.0.0",8888,8888

OK
+STA_CONNECTED:"e8:84:a5:8e:60:8e"
+DIST_STA_IP:"e8:84:a5:8e:60:8e","192.168.4.2"

+IPD,11:MESSAGE_0


+IPD,11:MESSAGE_1


+IPD,11:MESSAGE_2
Here's the logs from the stm32, now it waits for the sta to connect to the ap:
Library initialized!
Device reset detected!
Device reset detected!
Device reset detected!
ESP set to access-point-only mode
Global Event no. 14
Access point configured!
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14
New station connected to access point with MAC address: E8:84:A5:8E:60:8E

IP 192.168.4.2
 assigned to station with MAC address: E8:84:A5:8E:60:8E

Failed to start connection, error no. 2
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14
Global Event no. 14

**Reply by MaJerle** [reply]
You should always wait for station to connect, then start connection. You are doing the opposite. This may, or may not work.

**Reply by tommasoclini** [reply]
But as you can see in the logs, I made it wait for the STA to connect and then start the connection

**Reply by MaJerle** [reply]
You have not:

```
AT+CIPSTART="UDP","0.0.0.0",8888,8888
AT+CIPSTART="UDP","0.0.0.0",8888,8888
+LINK_CONN:0,0,"UDP",0,"0.0.0.0",8888,8888

OK
+STA_CONNECTED:"e8:84:a5:8e:60:8e"
+DIST_STA_IP:"e8:84:a5:8e:60:8e","192.168.4.2"

+IPD,11:MESSAGE_0
```

**Reply by tommasoclini** [reply]
I'm sorry, I meant the stm32 logs, in the logs you are referring I did like this to demonstrare that the esp is capable of receiving even if the connection is started before the station connects and the ip is not valid.
In the first logs I am manually sending commands with an usb to uart converted, in the second case, it's the stm32 that is susing lwesp

**Comment by tommasoclini** [comment]
So, how should I start a UDP transmission when the esp32 is the AP itself? In the comment where I showed the logs, they were two separate logs, as described in this comment:
> I'm sorry, I meant the stm32 logs, in the logs you are referring I did like this to demonstrare that the esp is capable of receiving even if the connection is started before the station connects and the ip is not valid. In the first logs I am manually sending commands with an usb to uart converted, in the second case, it's the stm32 that is susing lwesp
